doc/ECMP/Class-Based-Forwarding.md

```diff
@@ -0,0 +1,204 @@
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_NEXT_HOP, <b>SAI_OBJECT_TYPE_NEXT_HOP_GROUP</b>
+     */
+    SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_ID,
+```
+
+*Note: While this would also be a means to configure hierarchical ECMP, hierarchical ECMP is outside the scope of this proposal.*
+
+The forwarding-class for a packet may be selected via qos-map or ACL.
+
+```
+typedef  enum  _sai_qos_map_type_t
+{
+...
+    /** QOS Map to set DSCP to Forwarding class */
+    SAI_QOS_MAP_TYPE_DSCP_TO_FORWARDING_CLASS = 0x0000000d,
+
+    /** QOS Map to set EXP to Forwarding class */
+    SAI_QOS_MAP_TYPE_MPLS_EXP_TO_FORWARDING_CLASS = 0x0000000e,
+...
+} sai_qos_map_type_t;
+```
+
+```
+typedef enum _sai_acl_entry_attr_t
+...
+    /**
+     *  @brief  Set Forwarding Class
+     *
+     *  @type  sai_acl_action_data_t  sai_uint8_t
+     *  @flags  CREATE_AND_SET
+     *  @default  disabled
+     */
+    SAI_ACL_ENTRY_ATTR_ACTION_SET_FORWARDING_CLASS,
+...
+}  sai_acl_entry_attr_t;
+```
+
+If the packet is not assigned a forwarding-class, then the forwarding-class of the packet shall be 0. For example, if no qos-map or ACL is configured.
+
+*Resource monitoring considerations:*
+
+SAI_SWITCH_ATTR_MAX_NUMBER_OF_FORWARDING_CLASSES may be used to identify the maximum forwarding-class allowed.
+
+The SAI_OBJECT_TYPE_NEXT_HOP_GROUP_MAP object may be a limited resource. The sai_object_type_get_availability() API may be used to query the maximum number of permitted maps.
+
+*Class-based forwarding group configuration example:*
+```
+/******************************************************
+ * Create a forwarding-class -> index map.
+ * In this example, a simple 1:1 mapping is configured.
+ ******************************************************/
+ const int num_forwarding_classes = 8;
+
+ sai_object_id_t nh_group_map;
+
+ sai_map_t fc_map[num_forwarding_classes];
+ for (int i = 0; i < num_forwarding_classes; ++i) {
+     fc_map[i].key = i;
+     fc_map[i].value = i;
+ }
+
+ sai_map_list_t fc_map_list;
+ fc_map_list.key.count = num_forwarding_classes;
+ fc_map_list.key.list = fc_map;
+
+ attr.type = SAI_NEXT_HOP_GROUP_MAP_ATTR_TYPE;
+ attr.value.u32 = SAI_NEXT_HOP_GROUP_MAP_TYPE_FORWARDING_CLASS_TO_INDEX;
+ attrs.push_back(attr);
+
+ attr.id = SAI_NEXT_HOP_GROUP_MAP_ATTR_MAP_TO_VALUE_LIST;
+ attr.value.maplist = fc_map_list;
+ attrs.push_back(attr);
+
+ sai_next_hop_group_api->create_next_hop_group_map(
+     &nh_group_map,
+     g_switch_id,
+     attrs.size(),
+     attrs.data());
+
+ /*****************************************************
+  * Create a class-based forwarding group
+  *****************************************************/
+ attrs.clear();
+
+ sai_object_id_t nh_group;
+
+ attr.id = SAI_NEXT_HOP_GROUP_ATTR_TYPE;
+ attr.value.u32 = SAI_NEXT_HOP_GROUP_TYPE_CLASS_BASED;
+ attrs.push_back(attr);
+
+ attr.id = SAI_NEXT_HOP_GROUP_ATTR_CONFIGURED_SIZE;
+ attr.value.u32 = num_forwarding_classes;
```

Comment:
Could we change the `num_forwarding_classes` to something else that would make it more obvious that is rather the number of members that we have to fill in, not the number of entries in the map?

At a first read I was confused as why that SAI attribute was reused like that before I remembered the example has a 1-1 mapping. Other first time readers might be confused by this as well.

Also, because of this, I think a better example would contain a mapping between `X` FCs and `Y` NHs, where `X > Y` just so we make it obvious that that is possible.

Reply:
Also, the `@validonly` flag of `SAI_NEXT_HOP_GROUP_ATTR_CONFIGURED_SIZE` states `SAI_NEXT_HOP_GROUP_ATTR_TYPE == SAI_NEXT_HOP_GROUP_TYPE_FINE_GRAIN_ECMP`. So either we need to update the attribute's comment, or it was miss-used here.